[Woo POS] Retrieve order remotely in payment collection instead of relying on existing order in storage

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12869 

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

### Why

Previously, we encountered a bug where the order amount in payment collection became outdated after the first checkout p1718989636432309/1718983933.798199-slack-C070SJRA8DP. The reason was that CollectOrderPaymentUseCase dispatches `OrderAction.retrieveOrder` which uses the existing order in storage if its modified date field hasn't changed remotely (which is the case for edited auto-draft orders at least).

To fix this, originally I upserted the synced order to storage in POSOrderService. But it has a side effect of auto-draft orders appearing in the orders tab until the next sync. After @joshheald's clarification, using the order from storage was not the intended behavior and we should always load the order remotely for payment validation.

### How

A new action `OrderAction.retrieveOrderRemotely` was added with some test cases (and I got stuck in the last test case from not saving the context before dispatching the action 😅 ).

Then, `CollectOrderPaymentUseCase` replaces `OrderAction.retrieveOrder` with `OrderAction.retrieveOrderRemotely` to always load the error from the network. A new error `CollectOrderPaymentUseCaseError.orderTotalChanged` is now thrown if the order total from the remote is different from the original order in the use case. The merchant is able to retry, and  `OrderAction.retrieveOrder` will be triggered again in case the merchant can fix this in some way. Please feel free to suggest handling the error in a better way.

Because of the change to use the order from a remote sync, the storage code in `POSOrderService` was removed completely.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Ensure that the previous issue p1718989636432309/1718983933.798199-slack-C070SJRA8DP that requires upserting the order does not occur anymore.

Prerequisite: the store is eligible for POS and the feature switch is enabled in Menu > Settings > Experimental Features > POS
- Launch app
- Go to Menu > Point of Sale
- Connect the reader
- Add something to the cart
- Go to checkout
- Wait for it to be ready to tap
- Go back to Add More
- Add another item
- Go to checkout
- Tap card when it's ready --> the payment should succeed in the end, and the corresponding order in the orders tab should have the same amount as in the order total

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

@jaclync also tested going through several transactions and going back and forth before the final payment.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

The error case looks like this:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/d48e7402-b315-454f-852b-67904c6a76cd" width="300" />

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Add `OrderAction.retrieveOrderRemotely` with unit tests.

jaclync committed:
Always retrieve order remotely for payment validation, and throw an error when the order total has changed.

jaclync committed:
Remove upserting synced order to storage from POSOrderService.

jaclync committed:
Remove storage references in POSOrderService.

jaclync committed:
Update order total localizable string.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Fix unit test build failures.

joshheald committed:
12869 Make orderTotalChanged non-retryable

Before they retry with this error, the merchant needs to go back and check they agree with the order items. A retry button should (essentially) never work.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CollectOrderPaymentUseCase.swift:
```diff
@@ -822,7 +824,7 @@ extension CollectOrderPaymentUseCaseError: CardPaymentErrorProtocol {
         case .paymentGatewayNotFound:
             return .restart
-        case .unknownErrorRefreshingOrder, .couldNotRefreshOrder:
+        case .couldNotRefreshOrder, .orderTotalChanged:
```
I don't think we should make this retryable, as the `Try again` button will always fail – at least in the existing app, probably in POS too though I wasn't able to recreate it there.

They need to go back to the order and refresh, then pay again.

I've updated this in 586568e

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/211">19.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald reviewed (approved):
Looks great!

I tested the existing app as well:

1. Navigate to `Orders > +`
2. Add something to the order
3. Tap `Create`
4. Edit the order in WP-Admin to change the total
5. In the app, tap `Collect Order Payment`
6. Select `Card Reader`
7. Observe that after connecting the reader, the payment fails with a non-retryable error saying the total has changed.


I changed the error to be non-retryable, and the error message to describe the merchant's next steps. In the existing app, the order is automatically updated in the view behind, because of the `retrieveOrderRemotely` call updating storage... however, in POS they would probably need to tap `Add More` and then `Checkout` again. I've not described that though.

As a future improvement for POS, we could also refresh the order somehow, so they don't have to go back and forth. This might be something we just do for all errors... but we can handle that separately

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CollectOrderPaymentUseCase.swift:
```diff
@@ -749,10 +751,10 @@ enum CollectOrderPaymentUseCaseError: LocalizedError {
-            "Unable to process payment. We could not fetch the latest order details. Please check your network " +
-            "connection and try again.",
-            comment: "Error message when collecting an In-Person Payment and unable to update the order.")
+        static let orderTotalChangedLocalizedDescription = NSLocalizedString(
+            "collectOrderPaymentUseCase.error.message.orderTotalChanged",
+            value: "Unable to process payment. Order total has changed since the beginning of payment.",
+            comment: "Error message when collecting an In-Person Payment and the order total has changed remotely.")
```
```suggestion
        static let orderTotalChangedLocalizedDescription = NSLocalizedString(
            "collectOrderPaymentUseCase.error.message.orderTotalChanged",
            value: "Order total has changed since the beginning of payment. Please go back and check the order is " +
            "correct, then try the payment again.",
            comment: "Error message when collecting an In-Person Payment and the order total has changed remotely.")
```

joshheald committed:
12869 Update error message to explain next steps
